DepClean Gradle Plugin implementation. (Not for merge)

ABHAY0O7 opened the pull request:
Currently, I think I have implemented the core logic correctly, but still, there are some drawbacks that I am still trying to improve.
Also, the code that I have written until now is very unmanaged, there are lots of unnecessary commented out code, there are lots of method that I used for debugging and I provided very few java docs, so sorry for that. Please ignore these parts of code.
Currently to test the plugin,

1. First, Run `gradle clean build` without changing anything,
2. Run `gradle publishToMavenLocal`
3. Now uncomment the buildscript and apply plugin part that I have commented in build.gradle and again run `gradle clean build`
4. Run `gradle copyDependenciesLocally` task.
5. Run final `gradle debloat` task and you can see the results.

Thank you

ABHAY0O7 commented:
Hello @tdurieux @bbaudry 
I was thinking that if we could have a meet this weekend or may after that so that I can discuss the current status of the plugin, and also some problems that I am facing right now. The problems are quite complicated which I think can only be illustrated better at a meet, so it will be better for me.

tdurieux reviewed (changes requested):
Could you remove all the Gradle folders, we would like to avoid to have .ar files inside the project?

Could you remove all the .gradle folder from the test resources?

Could you rename the folder depclean-gradle-plugin/src/Test to test?

It is for the moment difficult to review the changes when so many files are included in the PR. After you remove the folders I will continue to review the PR

ABHAY0O7 commented:
> Could you remove all the .gradle folder from the test resources?

I already removed them but I forgot to remove their caches, but now I removed them too.

> Could you rename the folder depclean-gradle-plugin/src/Test to test?

Renamed 👍🏽 

> Could you remove all the Gradle folders, we would like to avoid to have .ar files inside the project?

Are you talking about gradle.wrapper folders? If no, then please can you be more specific but If yes then in my knowledge we should not involve these folders in .gitignore as they are important ones. But if you want these folders to be ignored for now then I can remove them too. So please can you be more specific about this question that what I have to do.

tdurieux commented on depclean-gradle-plugin/build.gradle:
```diff
@@ -0,0 +1,50 @@
+//buildscript{
```
Can you remove the comments?

ABHAY0O7 replied:
Ok.

tdurieux replied:
you uncommented, you should remove those lines. if the plugin is not present on the system, it is not possible to compile it.

ABHAY0O7 replied:
Sorry that was my mistake.
I will correct it in my next push.

tdurieux commented on depclean-gradle-plugin/gradle/wrapper/gradle-wrapper.properties:
```diff
@@ -0,0 +1,5 @@
+distributionBase=GRADLE_USER_HOME
```
Do you have to keep the wrapper folder?

ABHAY0O7 replied:
Yes, it will be better if we keep the wrapper folder as it is very useful. In case if someone doesn't have Gradle installed in their system or is unupdated with the latest version then this wrapper folder will help them to build the project without any failure. For more information visit [here](https://stackoverflow.com/a/20351077/15557529)

tdurieux commented on depclean-gradle-plugin/src/main/java/se/kth/depclean/DepCleanGradleTask.java:
```diff
@@ -0,0 +1,23 @@
+   * Action that analyzes the project dependencies and show the results.
+   */
+  @TaskAction
+  public void printDependencies() {
```
The name of the function is not related to its body

ABHAY0O7 replied:
Can it be analyzeProject() or analyzeGradleProject()?

tdurieux commented on depclean-gradle-plugin/src/main/java/se/kth/depclean/DepCleanGradleTask.java:
```diff
@@ -0,0 +1,23 @@
+    Action<Project> defaultAction = new DepCleanGradleAction();
+    getProject().allprojects(defaultAction);
+
+    getProject().getLogger().lifecycle("\nTask execution ends.");
```
Can be removed or should be improved.

ABHAY0O7 replied:
Better remove.

tdurieux commented on depclean-gradle-plugin/src/main/java/se/kth/depclean/analysis/GradleProjectDependencyAnalysis.java:
```diff
@@ -0,0 +1,144 @@
+  private final Set<ResolvedArtifact> unusedDeclaredArtifacts;
+
+  /**
+   * Ctor.
```
to remove

ABHAY0O7 replied:
Are you talking about that comment here?

tdurieux commented on depclean-gradle-plugin/src/main/java/se/kth/depclean/analysis/GradleProjectDependencyAnalysis.java:
```diff
@@ -0,0 +1,144 @@
+//   */
+//  public GradleProjectDependencyAnalysis ignoreNonCompile() {
+//    Set<ResolvedArtifact> filteredUnusedDeclared = new HashSet<>(unusedDeclaredArtifacts);
+////    filteredUnusedDeclared.removeIf(resolvedArtifact -> !resolvedArtifact.getScope().equals(Artifact.SCOPE_COMPILE));
```
Will it be used?

ABHAY0O7 replied:
No, it will be of no use as I ignore scopes using the coordinates of the artifacts and not with any Artifact object itself.

ABHAY0O7 replied:
So, I will remove it.

tdurieux commented on depclean-gradle-plugin/src/test/resources/all_dependencies_used/gradle/wrapper/gradle-wrapper.properties:
```diff
@@ -0,0 +1,5 @@
+distributionBase=GRADLE_USER_HOME
```
remove wrapper folder.

ABHAY0O7 replied:
Ya, I think we can remove these resources wrapper folders.

tdurieux commented on depclean-gradle-plugin/src/main/java/se/kth/depclean/DepCleanGradleAction.java:
```diff
@@ -0,0 +1,506 @@
+
+  @SneakyThrows
+  @Override
+  public void execute(@NotNull Project project) {
```
This method is too big, could you refactor it and split it into smaller entities (to improve maintainability)?

ABHAY0O7 replied:
Yes, I will try to split it as more as possible.

tdurieux commented on depclean-gradle-plugin/src/main/java/se/kth/depclean/analysis/DefaultGradleProjectDependencyAnalyzer.java:
```diff
@@ -0,0 +1,331 @@
+  /**
+   * If true, the project's classes in target/test-classes are not going to be analyzed.
+   */
+  private final boolean isIgnoredTest;
```
Could you create the same parameters that we have in the maven plugins? (could be done in a separate PR)

ABHAY0O7 replied:
I added all the parameters that we have in the maven plugin except for 3 parameters about which I talked about [here](https://github.com/castor-software/depclean/issues/92#issuecomment-873566633)
Please take a look there and then reply accordingly.

tdurieux replied:
I would still implement those parameters, we can have set the default to false for those additional parameters but those are useful.

tdurieux reviewed (changes requested):
Good job, the PR seems good. There are few elements that need to improve but it should be relatively fast.

ABHAY0O7 commented on depclean-gradle-plugin/src/main/java/se/kth/depclean/depcleanGradleAction.java:
```diff
@@ -0,0 +1,304 @@
+import se.kth.depclean.core.analysis.ProjectDependencyAnalyzerException;
+import se.kth.depclean.util.JarUtils;
+
+public class depcleanGradleAction implements Action<Project> {
```
Ok.

ABHAY0O7 commented:
Hello @tdurieux 
I have implemented all the requested changes except for those 2 which are still unresolved.
To resolve them, you have to be more clear that what I have to do.

Thank you.

tdurieux commented on depclean-gradle-plugin/src/main/java/se/kth/depclean/DepCleanGradleAction.java:
```diff
@@ -0,0 +1,852 @@
+    final String projectDirPath = project.getProjectDir().getAbsolutePath() + Sep;
+
+    // Path to the dependency directory.
+    final String dependencyDirPath = projectDirPath + "build" + Sep + "Dependency";
```
You could use Paths.get: https://docs.oracle.com/javase/7/docs/api/java/nio/file/Paths.html#get(java.lang.String,%20java.lang.String...)

ABHAY0O7 replied:
Thanks. that's a nice refactoring. I will implement it (wherever possible) in my next push.

tdurieux commented on depclean-gradle-plugin/src/main/java/se/kth/depclean/analysis/DefaultGradleProjectDependencyAnalyzer.java:
```diff
@@ -0,0 +1,345 @@
+            + sep + "classes";
+
+    // Analyze src classes in the project
+    File outputJavaDir = new File(classesDir + sep + "java" + sep + "main");
```
You could use Paths.get: https://docs.oracle.com/javase/7/docs/api/java/nio/file/Paths.html#get(java.lang.String,%20java.lang.String...)

tdurieux commented:
Looks good. `depclean-gradle-plugin/src/main/java/se/kth/depclean/DepCleanGradleAction.java` is really big maybe we should work on a way to make it smaller and easier to maintain.

ABHAY0O7 commented:
> Looks good. `depclean-gradle-plugin/src/main/java/se/kth/depclean/DepCleanGradleAction.java` is really big maybe we should work on a way to make it smaller and easier to maintain.

Hii @tdurieux,
As you suggested I reduced the size of the `DepCleanGradleAction`, to do this I introduce two new utility classes.
I also used the `Path` object instead of `String` path which saves a little bit extra lines of code.

tdurieux commented:
I just realized that Github Action does not execute the test from the gradle plugin. Could you update https://github.com/castor-software/depclean/blob/master/.github/workflows/build.yml to include the gradle plugin tests. 
I think it is the final task before I can merge the PR.

It is a good job, continue like this!

ABHAY0O7 commented:
Hii @tdurieux 
I want you to please don't merge this PR for now.
This is PR is very messy and there are some useless commits also, so to make the history very clean and easily understandable, give me some time, I will update you soon.

Thank you

ABHAY0O7 commented:
Hii @tdurieux 
I have started to make the git history clean and understandable.
I have opened some PR's with their order of merge in their titles, now the commits are more clear and understandable with some description. 
I want you to review and merge the PRs as soon as possible, so that when I push new commits to my branches then they should be on new PR. :))

Thank you
Abhay

ABHAY0O7 commented:
Hii @tdurieux 
Nice to see you :))
I will update you with remaining PRs up to 7 PM GMT.

Thank you

ABHAY0O7 commented:
The work of this PR is divided and implemented int these PRs :

- #94 
- #95 
- #96 
- #97 
- #98 
- #99
